### ISSUE TITLE
Setup Julia Kernel on the Docker version of Cocalc

### ISSUE THREAD

**Issue by F-Forget** [sagemathinc/cocalc-docker#43]
Hi,

I’ve downloaded the latest version of Cocalc from Docker Hub and I actually want to host a Cocalc server on my computer. I actually want to add Julia 1.0.3 has a new kernel. I’ve followed Marcelo Forets (https://github.com/mforets) tutorial about adding Julia (https://github.com/sagemathinc/cocalc/wiki/JuliaCustom), combined with the information given in the Cocalc Dockerfile : 

`# Install Julia`
`RUN cd /tmp \`
 `&& wget https://julialang-s3.julialang.org/bin/linux/x64/1.0/julia-1.0.3-linux-x86_64.tar.gz \`
 `&& tar xf julia-1.0.3-linux-x86_64.tar.gz -C /opt \`
` && rm  -f julia-1.0.3-linux-x86_64.tar.gz \`
` && mv /opt/julia-* /opt/julia \`
` && ln -s /opt/julia/bin/julia /usr/local/bin`

`# Install IJulia kernel`
`RUN echo '\`
`ENV["JUPYTER"] = "/usr/local/bin/jupyter"; \`
`ENV["JULIA_PKGDIR"] = "/opt/julia/share/julia/site"; \`
`Pkg.init(); \`
`Pkg.add("IJulia");' | julia \`
` && mv -i "$HOME/.local/share/jupyter/kernels/julia-1.0" "/usr/local/share/jupyter/kernels/"`

But wasn’t able to make it work.

Using “jupyter kernelspec list” command, I was able to see Julia : 

<img width="477" alt="Capture d’écran, le 2019-04-09 à 18 02 09" src="https://user-images.githubusercontent.com/7460096/55838285-e52c1380-5af1-11e9-83df-5c090ae2679c.png">

But it did not successfully appear in Cocalc Kernel lists : 

<img width="275" alt="Capture d’écran, le 2019-04-09 à 18 02 24" src="https://user-images.githubusercontent.com/7460096/55838289-ea895e00-5af1-11e9-89fd-d7c14cb06b6a.png">

How can I make it appear?

Thanks!

Francis Forget

**Comment by haraldschilly** [comment]
At the bottom of that kernel menu is  a "refresh kernel list" entry. It's basically a matter of updating the frontend's knowledge and this facilitates it. Equally well you can refresh the tab in your browser, which clears all that prior knowledge and causes that list to be loaded fresh as well.

**Comment by F-Forget** [comment]
I’ve actually tried both options! Refresh the list, refresh the browser and also restart the containers! 

None of those options works! 

Please, can you try to make it work on your side and tell me a way to make it work?

I’m really not sure what else I can do to make it appear in the list :( 

Thank you!

**Comment by haraldschilly** [comment]
Hmm, I'm not sure what's going on. I found traces of Julia and IJulia in the [Dockerfile of this repo here](https://github.com/sagemathinc/cocalc-docker/blob/master/Dockerfile#L212) ... maybe it's best to update it to 1.0.3 and see what's going on? I bet there is no interest in Julia before 1.0. E.g. I wondering if `using Pkg` is required to make the `Pkg....` commands work.

Did you try to run the kernel in a terminal? i.e. `jupyter console --kernel=julia-1.0`

If that works, we're at least know that the jupyter<->julia communication works.

**Comment by F-Forget** [comment]
I've tried to run the kernel in the terminal and it worked fine using the command you posted above : `jupyter console --kernel=julia-1.0`

I was also able to make the Pkg package work. I've used `using Pkg` and `Pkg.add("IJulia")` successfully without errors. I feel like CoCalc is just not able to see Julia in the kernel list, but that everything else is set up properly! 

Again I could be wrong and I hope you can help me with that! 😄

**Comment by F-Forget** [comment]
I've also saw that there is traces of Julia in the Dockerfile, but instead of re-building the CoCalc image with those lines enabled (which failed because a Sage modules failed to install properly due to the unavailability of giac-1.5.0.37.p0). That's why I decided to get the image using 
`docker run --name=cocalc -d -v ~/cocalc:/projects -p 443:443 sagemathinc/cocalc`
and then installed Julia using the Dockerfile steps and the tutorial I posted above!

**Comment by williamstein** [comment]
> I've tried to run the kernel in the terminal and it worked fine using the command you posted above : jupyter console --kernel=julia-1.0

I think you're testing and installing in the wrong terminal.   You're doing that as the root user in the container, but you need to ensure that julia is fully usable as any normal user with all the right permissions.  

E.g., try +New --> Terminal from the cocalc interface, then try using Julia from **that** command line.  Does it work?

**Comment by F-Forget** [comment]
I’ll need to test that cause that’s a very good point @williamstein! Can you please tell me how can I install it for all users in order to get CoCalc to actually see it? 

Again, I don’t want to rebuild the entire image since I get an error related to the unavailability of giac-1.5.0.37.p0…

Is there an actual procedure on how to add a new kernel and actually made it appear in CoCalc?

**Comment by F-Forget** [comment]
I ended up tweeting it with what you said! 

I actually created a Terminal for a project, and manage to run the necessary command in order for my project to have the Julia Kernel. I basically went back where my root user had Julia, and ran the following commands : 😊 

`using Pkg`  
`Pkg.add("IJulia")`

Then I was able to access the Julia Kernel from CoCalc! 

Thank you for you help! 

I do have another question and I'm not sure if it's the right place to ask it, but I want to be able to access my localhost from my entire local network. How could I do that? 

Thank you 😄
